addons/knowledge/controllers/main.py

```diff
@@ -0,0 +1,319 @@
+                member_values['email'] = member.partner_id.email
+            members_values.append(member_values)
+        internal_permission_field = request.env['knowledge.article']._fields['internal_permission']
+        permission_field = request.env['knowledge.article.member']._fields['permission']
+        user_is_admin = request.env.user.has_group('base.group_system')
+        return {
+            'internal_permission_options': internal_permission_field.get_description(request.env).get('selection', []),
+            'internal_permission': article.inherited_permission,
+            'parent_permission': article.parent_id.inherited_permission,
+            'based_on': article.inherited_permission_parent_id.display_name,
+            'based_on_id': article.inherited_permission_parent_id.id,
+            'members_options': permission_field.get_description(request.env).get('selection', []),
+            'members': members_values,
+            'is_sync': is_sync,
+            'parent_id': article.parent_id.id,
+            'parent_name': article.parent_id.display_name,
+            'user_is_admin': user_is_admin,
+            'show_admin_tip': user_is_admin and article.user_permission != 'write',
+        }
+
+    @http.route('/knowledge/article/set_member_permission', type='json', auth='user')
+    def article_set_member_permission(self, article_id, permission, member_id=False, partner_id=False, **post):
+        """
+        Sets the permission of the given member for the given article.
+        **Note**: The user needs "write" permission to change the permission of a user.
+        :param article_id: (int) article id
+        :param permission: (string) permission
+        :param member_id: (int) member id
+        :param partner_id: (int) partner id. Typically given when the member is based on a parent article.
+        """
+        article = request.env['knowledge.article'].sudo().browse(article_id)
+        if not article.user_can_write:
+            return werkzeug.exceptions.Forbidden()
+        previous_category = article.category
+        if partner_id:  # If based on
+            partner = request.env['res.partner'].browse(partner_id).exists()
+            if not article or not partner:
+                return {'success': False}
+            member = request.env['knowledge.article.member'].browse(member_id).exists()
+            downgrade = bool(member) and member.permission > permission
+            if (not downgrade and not article.invite_members(partner, permission, send_mail=False)) \
+                    or (downgrade and not article._desync_access_from_parents(partner.ids, permission)):
+                return {'success': False}
+        elif not article or not article._set_member_permission(member_id, permission):
+            return {'success': False}
+        result = {'success': True}
+        if article.category != previous_category:
+            result['reload_tree'] = True
+        return result
```

Comment:
Instead of "partner_id", we should pass "based_on".
Passing partner_id makes it complicated to follow because every member has a partner_id, it should not define the use case of "based on vs existing member".

Sugg:

```py
result = {'success': True}

article = request.env['knowledge.article'].browse(article_id)  # -> maybe not necessary if we change the method to contain a <model> in the route
member = request.env['knowledge.article.member'].browse(member_id).exists()
based_on_article = request.env['knowledge.article'].browse(based_on_article_id).exists()

if not article.user_can_write:
    return werkzeug.exceptions.Forbidden()

previous_category = article.category

if based_on_article:
    downgrade = bool(member) and member.permission > permission
    if member and downgrade:
        # we downgrade the permission of a member from an inherited permission -> desync from the parent
        article._desync_access_from_parents(member.partner_id, permission)
    elif member:
        # upgrade the member permission (but don't desync)
        member.sudo().write({'permission': permission})
    else:
       # upgrade the permission but no member yet... not sure when that happens?
       pass
else:
    member.sudo().write({'permission': permission})

if article.category != previous_category:
    result['reload_tree'] = True

return result
```

When writing the suggestion, it seems like we are handling the case when there is no member and a partner_id.
Is that possible? You would think this is the inviting flow?